### ISSUE TITLE
Type inference corner case in combination of object spread with destructuring

### ISSUE THREAD

**Issue by Kovensky** [Microsoft/TypeScript#12532]
**TypeScript Version:**  2.2.0-dev.20161127

**Code**

```ts
declare var keyPrototype: any
declare function getKeyData(): Promise<{ key: any }>

async function getKey () {
  const { key } = {
    key: {
      ...keyPrototype
    },
    ...await getKeyData()
  }
  return key
}
```

**Expected behavior:**

Code to typecheck. The resulting `key` const should be a shallow copy of `keyPrototype`, with certain keys overridden by the result of `getKeyData`.

**Actual behavior:**

Error on the `key` declaration inside `const { key }`: "Initializer provides no value for this binding element and the binding element has no default value", despite the RHS being guaranteed to always have a `key` key.

**Comment by DarylCantrell** [comment]
This is probably a duplicate. In TSX file, this works fine:

    const { children, ...rest} = this.props;  // OK

When I try to do this, I get the same error you're seeing:

    const { children, ...rest} = {    // Error
        ...this.props,
        someProp: forcedValue,
    }

Works if I supply a default value, but I don't think it should require one:

    const { children = null, ...rest} = {
        ...this.props,
        prop: forcedValue,
    }

**Comment by robyoder** [comment]
I'm having this issue as well.

**Comment by robyoder** [comment]
This appears to be fixed in version `2.7.0-dev.20171230`. Perhaps by #19513? Recommend for closing.

**Comment by sandersn** [comment]
Yep, original repro works now.
